Nested use of same SVG resource fails

rniwa opened the pull request:
#### 99265dce4d2f8db13b22db2f9018e15cd666bc03
<pre>
Nested use of same SVG resource fails
<a href="https://bugs.webkit.org/show_bug.cgi?id=130127">https://bugs.webkit.org/show_bug.cgi?id=130127</a>

Reviewed by Said Abou-Hallawa.

Merge <a href="https://src.chromium.org/viewvc/blink?view=revision&amp">https://src.chromium.org/viewvc/blink?view=revision&amp</a>;revision=175129

* LayoutTests/svg/custom/pattern-3-step-cycle-dynamic-1-expected.txt: Added.
* LayoutTests/svg/custom/pattern-3-step-cycle-dynamic-1.html: Added.
* LayoutTests/svg/custom/pattern-3-step-cycle-dynamic-2-expected.txt: Added.
* LayoutTests/svg/custom/pattern-3-step-cycle-dynamic-2.html: Added.
* LayoutTests/svg/custom/pattern-3-step-cycle-dynamic-3-expected.txt: Added.
* LayoutTests/svg/custom/pattern-3-step-cycle-dynamic-3.html: Added.
* LayoutTests/svg/custom/pattern-3-step-cycle-dynamic-4-expected.txt: Added.
* LayoutTests/svg/custom/pattern-3-step-cycle-dynamic-4.html: Added.
* LayoutTests/svg/custom/pattern-directly-and-indirectly-referenced-expected.svg: Added.
* LayoutTests/svg/custom/pattern-directly-and-indirectly-referenced.svg: Added.
* LayoutTests/svg/custom/pattern-false-resource-cycle-expected.html: Added.
* LayoutTests/svg/custom/pattern-false-resource-cycle.html: Added.
* LayoutTests/svg/custom/pattern-within-other-pattern-expected.html: Added.
* LayoutTests/svg/custom/pattern-within-other-pattern.html: Added.
* Source/WebCore/rendering/svg/SVGResourcesCycleSolver.cpp:
(WebCore::SVGResourcesCycleSolver::SVGResourcesCycleSolver): Deleted.
(WebCore::SVGResourcesCycleSolver::resourceContainsCycles):
(WebCore::SVGResourcesCycleSolver::resolveCycles):
(WebCore::SVGResourcesCycleSolver::breakCycle):
* Source/WebCore/rendering/svg/SVGResourcesCycleSolver.h:
(WebCore::SVGResourcesCycleSolver::SVGResourcesCycleSolver):

Canonical link: <a href="https://commits.webkit.org/264618@main">https://commits.webkit.org/264618@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a59a590da68c9f52166661fc4b07fafe58dd5c41

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8085 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8372 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8586 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9740 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8181 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8090 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10358 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8283 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11039 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8229 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9314 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7365 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9861 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6615 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7402 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14972 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7738 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7527 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10879 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7995 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6504 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7305 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/7315 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11514 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/970 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7737 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/rendering/svg/SVGResourcesCycleSolver.cpp:
```diff
@@ -50,134 +41,75 @@ SVGResourcesCycleSolver::SVGResourcesCycleSolver(RenderElement& renderer, SVGRes
     for (auto& resource : localResources) {
-        if (ancestorResources.contains(resource) || resourceContainsCycles(resource)) {
-            LOG_DEBUG_CYCLE("\n**** Detected a cycle (see the last test in the output above) ****\n");
+        if (m_activeResources.contains(resource) || resourceContainsCycles(resource))
```
When can `m_activeResources.contains(resource)` be true? `resourceContainsCycles()` makes sure it removes all the items it adds recursively. So  m_activeResources has to have zero or one entry outside `resourceContainsCycles()` depending on whether m_renderer was added at the tope of this function or not.

rniwa replied:
Right. `m_activeResources.contains(resource)` will only be true if there is a self reference.

shallawa commented on Source/WebCore/rendering/svg/SVGResourcesCycleSolver.cpp:
```diff
@@ -50,134 +41,75 @@ SVGResourcesCycleSolver::SVGResourcesCycleSolver(RenderElement& renderer, SVGRes
-            LOG_DEBUG_CYCLE("(%p) Child %p had cached resources. Check them.\n", &renderer, &child);
+bool SVGResourcesCycleSolver::resourceContainsCycles(RenderSVGResourceContainer& resource)
+{
+    if (m_acyclicResources.contains(resource))
```
Why do we check the `m_acyclicResources` only? And what is the definition of `acyclicResources` here since we `breakCycle()` for the cyclic ones. I think we should check `allVisitedResources` instead.

rniwa replied:
This is an early return for when we had already determined that resource doesn't have a cycle.
Without this check, we can end up checking the same resource doesn't have a cycle over & over.

shallawa commented on Source/WebCore/rendering/svg/SVGResourcesCycleSolver.cpp:
```diff
@@ -50,134 +41,75 @@ SVGResourcesCycleSolver::SVGResourcesCycleSolver(RenderElement& renderer, SVGRes
 
-    LOG_DEBUG_CYCLE("\n(%p) No cycles found\n", &renderer);
-
+    m_acyclicResources.add(resource);
```
I think the caller of this function should make this add.

rniwa replied:
Hm.. that would mean we'd have to duplicate this code in two places though.

shallawa commented on Source/WebCore/rendering/svg/SVGResourcesCycleSolver.cpp:
```diff
@@ -50,134 +41,75 @@ SVGResourcesCycleSolver::SVGResourcesCycleSolver(RenderElement& renderer, SVGRes
 void SVGResourcesCycleSolver::resolveCycles()
 {
-    ASSERT(m_allResources.isEmptyIgnoringNullReferences());
+    ASSERT(m_activeResources.isEmptyIgnoringNullReferences());
```
The life cycle of this class suggests that `m_activeResources` can be just a local variable. So we do not have to check it is empty here and we do not have clear at the end of this function.

rniwa replied:
Yeah, that is quite true. Making this a static function makes sense to me.

shallawa commented on Source/WebCore/rendering/svg/SVGResourcesCycleSolver.h:
```diff
@@ -38,12 +38,13 @@ class SVGResourcesCycleSolver {
     SVGResources& m_resources;
-    WeakHashSet<RenderSVGResourceContainer> m_allResources; 
+    WeakHashSet<RenderSVGResourceContainer> m_activeResources;
+    WeakHashSet<RenderSVGResourceContainer> m_acyclicResources;
```
I would suggest making SVGResourcesCycleSolver no constructible class and making all its methods be static functions. So the caller can say:

```
    SVGResourcesCycleSolver::resolveCycles(renderer, resources);
```
instead of 

```
    SVGResourcesCycleSolver solver(renderer, resources);
    solver.resolveCycles();
```

All these members can be just arguments passed from one function to another.

rniwa replied:
Okay, done.

shallawa commented on Source/WebCore/rendering/svg/SVGResourcesCycleSolver.cpp:
```diff
@@ -50,134 +41,75 @@ SVGResourcesCycleSolver::SVGResourcesCycleSolver(RenderElement& renderer, SVGRes
-            // A child of the given 'resource' contains resources.
-            WeakHashSet<RenderSVGResourceContainer> childResourceSet;
-            childResources->buildSetOfResources(childResourceSet);
+    ActiveResourceAcquisition activeResourceAcquisition { m_activeResources, resource };
```
Can't we use ScopeExit instead:
```
    m_activeResources.add(resource);
    auto activeResourceAcquisition = WTF::makeScopeExit([&]() {
        m_activeResources.remove(resource);
    });
```

rniwa replied:
Sure, I can do that.

rniwa committed:
Nested use of same SVG resource fails
https://bugs.webkit.org/show_bug.cgi?id=130127

Reviewed by Said Abou-Hallawa.

Merge https://src.chromium.org/viewvc/blink?view=revision&revision=175129

* LayoutTests/svg/custom/pattern-3-step-cycle-dynamic-1-expected.txt: Added.
* LayoutTests/svg/custom/pattern-3-step-cycle-dynamic-1.html: Added.
* LayoutTests/svg/custom/pattern-3-step-cycle-dynamic-2-expected.txt: Added.
* LayoutTests/svg/custom/pattern-3-step-cycle-dynamic-2.html: Added.
* LayoutTests/svg/custom/pattern-3-step-cycle-dynamic-3-expected.txt: Added.
* LayoutTests/svg/custom/pattern-3-step-cycle-dynamic-3.html: Added.
* LayoutTests/svg/custom/pattern-3-step-cycle-dynamic-4-expected.txt: Added.
* LayoutTests/svg/custom/pattern-3-step-cycle-dynamic-4.html: Added.
* LayoutTests/svg/custom/pattern-directly-and-indirectly-referenced-expected.svg: Added.
* LayoutTests/svg/custom/pattern-directly-and-indirectly-referenced.svg: Added.
* LayoutTests/svg/custom/pattern-false-resource-cycle-expected.html: Added.
* LayoutTests/svg/custom/pattern-false-resource-cycle.html: Added.
* LayoutTests/svg/custom/pattern-within-other-pattern-expected.html: Added.
* LayoutTests/svg/custom/pattern-within-other-pattern.html: Added.
* Source/WebCore/rendering/svg/SVGResourcesCycleSolver.cpp:
(WebCore::SVGResourcesCycleSolver::SVGResourcesCycleSolver): Deleted.
(WebCore::SVGResourcesCycleSolver::resourceContainsCycles):
(WebCore::SVGResourcesCycleSolver::resolveCycles):
(WebCore::SVGResourcesCycleSolver::breakCycle):
* Source/WebCore/rendering/svg/SVGResourcesCycleSolver.h:
(WebCore::SVGResourcesCycleSolver::SVGResourcesCycleSolver):

Canonical link: https://commits.webkit.org/264618@main
